hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java

```diff
@@ -293,14 +293,26 @@ protected void updateIndex(Id ilId, HugeElement element, boolean removed) {
                 value = ConditionQuery.concatValues(allPropValues);
                 assert !value.equals("");
                 Id id = element.id();
-                // TODO: add lock for updating unique index
-                if (!removed && this.existUniqueValue(indexLabel, value, id)) {
-                    throw new IllegalArgumentException(String.format(
+
+                LockUtil.Locks locks = new LockUtil.Locks(this.graphName());
+                try {
+                    if (element.isVertex()) {
+                        locks.lockWrites(LockUtil.VERTEX_UPDATE, id);
+                    } else {
+                        locks.lockWrites(LockUtil.EDGE_UPDATE, id);
+                    }
+                    if (!removed &&
+                        this.existUniqueValue(indexLabel, value, id)) {
+                        throw new IllegalArgumentException(String.format(
                               "Unique constraint %s conflict is found for %s",
                               indexLabel, element));
+                    }
+                    this.updateIndex(indexLabel, value, element.id(),
+                                     expiredTime, removed);
+                } finally {
+                    locks.unlock();
```

Comment:
expect to release lock after tx commit, please reference `GraphTransaction.locksTable`